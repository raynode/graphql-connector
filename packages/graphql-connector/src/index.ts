export * from './base-schema-generator'
export * from './naming-strategy'
export * from './type-converter'
export * from './filter-field-generator'
export * from './model'
export * from './model-mapper'
export * from './utils'
export * from './configuration'
export * from './order-mapper'

