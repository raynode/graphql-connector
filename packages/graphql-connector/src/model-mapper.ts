import {
  AnyModel,
  Model,
  modelCreator,
  ModelResolver,
  PartialAssociation,
  PartialAssociations,
  PartialAttribute,
  PartialAttributes,
} from './model'

export type GeneratedModelMapper<Types, Models> = <Key extends keyof Models>(
  key: Key,
  model: Models[Key],
) => AnyModel<Types, Models, Models[Key]>

export const createModelMapper = <Types, Models>(
  mapper: <Key extends keyof Models>(
    model: Models[Key],
    addAttribute: <Inst>(attribute: PartialAttribute<Inst, Types>) => void,
    addAssociation: <Inst>(association: PartialAssociation<Inst, Models>) => void,
  ) => ModelResolver<any>,
): GeneratedModelMapper<Types, Models> => {
  const creator = modelCreator<Types, Models>()
  return <Key extends keyof Models, Inst>(key: Key, model: Models[Key]) => {
    const attributes: PartialAttributes<any, any, Types> = {}
    const associations: PartialAssociations<any, any, Models> = {}
    const reducers = mapper(
      model,
      attribute => (attributes[attribute.name] = attribute),
      association => (associations[association.name] = association),
    )
    return creator<Key, Inst>(key, attributes, associations, reducers, model)
  }
}
