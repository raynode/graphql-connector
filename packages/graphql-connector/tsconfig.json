{
  "compilerOptions": {
    "baseUrl": "./src",
    "declaration": true,
    "lib": ["es2016", "dom", "es7", "esnext"],
    "module": "commonjs",
    "moduleResolution": "node",
    "noEmitOnError": true,
    "noImplicitAny": false,
    "outDir": "lib",
    "rootDir": "src",
    "sourceMap": true,
    "target": "esnext"
  },
  "include": [
    "./src/**/*.ts"
  ]
}
