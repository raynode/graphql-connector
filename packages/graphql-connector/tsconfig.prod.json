{
  "extends": "./tsconfig.json",
  "exclude": [
    "src/**/*.spec.ts",
    "src/tests"
  ],
  "include": [
    "src"
  ]
}
