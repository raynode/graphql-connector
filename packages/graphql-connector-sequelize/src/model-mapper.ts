import { AnyModel, createModelMapper, GeneratedModelMapper, Page } from '@raynode/graphql-connector'
import { capitalize } from 'inflection'
import * as Sequelize from 'sequelize'
import { applyParser, createEmptyFilter, ParsedFilter, parser } from './filter-parser'
import { DataTypes } from './type-guards'

// tslint:disable-next-line:no-duplicate-imports
import { AssociationOptions, DefineAttributeColumnOptions } from 'sequelize'
declare module 'sequelize' {
  interface DefineAttributeColumnOptions {
    // should this attribute be represented in the SDL and be handled by graphql-connector (default: true)
    visible?: boolean
    // should this attribute be in the update mutation (default: true)
    updateable?: boolean
    // should this attribute be in the create mutation (default: true)
    createable?: boolean
  }

  interface AssociationOptions {
    pagination?: boolean
  }
}
// somehow the sequelize types are really really bad!
export interface SequelizeAttributeReference {
  model: string
  key: string
}

// export interface SequelizeAttribute extends Sequelize.DefineAttributeColumnOptions {
//   // should this attribute be represented in the SDL and be handled by graphql-connector (default: true)
//   visible?: boolean
//   // should this attribute be in the update mutation (default: true)
//   updateable?: boolean
//   // should this attribute be in the create mutation (default: true)
//   createable?: boolean
//   // these are correct, but incompatible... these sequelize types...
//   type: DataTypes | Sequelize.DataTypeAbstract
// }

// export type SequelizeAttributes<Keys extends string = string> = Record<Keys, string |Â SequelizeAttribute>

export type SequelizeAssociationTypes = 'HasOne' | 'HasMany' | 'BelongsTo' | 'BelongsToMany'

export interface SequelizeAssociation extends AssociationOptions {
  accessors: Record<string, string>
  as: string
  associationType: SequelizeAssociationTypes
  foreignKey: string
  foreignKeyAttribute: string
  identifier: string
  identifierField: string
  isSingleAssociation: boolean
  otherKey: string
  source: SequelizeModel
  target: SequelizeModel
  targetIdentifier: string
  targetKey: string
  targetKeyField: string
  targetKeyIsPrimary: string
  through?: { model: SequelizeModel, unique: boolean, scope: any }
}

export interface SequelizeModel extends Sequelize.Model<any, any> {
  rawAttributes: Record<string, DefineAttributeColumnOptions>
  associations: Record<string, SequelizeAssociation>
}

export interface Models {
  [x: string]: Sequelize.Model<any, any>
}

const createPage = (offset: number, limit: number, page: number): Page => ({
  offset, limit, page,
})

export const isListAssociation = (association: SequelizeAssociation) =>
  association.hasOwnProperty('isSingleAssociation') ? !association.isSingleAssociation :
    association.associationType === 'HasMany' || association.associationType === 'BelongsToMany'

export const modelMapper = createModelMapper<DataTypes, Models>((localModel, addAttribute, addAssociation) => {
  const model: SequelizeModel = localModel as any
  Object.keys(model.rawAttributes).forEach(name => {
    const attribute = model.rawAttributes[name]
    if(attribute.visible === false)
      return

    addAttribute({
      name,
      type: attribute.type as DataTypes,
      nonNull: attribute.allowNull === false || attribute.primaryKey === true,
      resolver: instance => instance[name],
    })
  })

  Object.keys(model.associations).forEach(name => {
    const association = model.associations[name]
    const list = isListAssociation(association)

    const targetKey = association.targetKey || association.foreignKey
    const through = association.through
    const attribute = association.target.rawAttributes[targetKey]
    const pagination = association.pagination

    const nonNull = !!list || !!through || attribute.allowNull

    addAssociation({
      name: association.as,
      model: association.target.name,
      list,
      nonNull,
      pagination,
      resolver: async (instance, args, context, info) => {
        // @TODO
        // this needs to correctly submit nodes and page when done!
        const res = await instance[association.accessors.get]()
        return list
        ? {
          nodes: res,
          page: createPage(0, 100, 0),
        } : res
      },
    })
  })

  const findAll = async (include, where, order?) => {
    const nodes = await model.findAll({ include, where, order })
    return {
      nodes,
      page: createPage(0, 100, 0),
    }
  }

  const findFromModel = (model: SequelizeModel, list: boolean, filter: ParsedFilter) => list
  ? model.findAll(filter)
  : model.findOne(filter)

  const getModelData = (association: SequelizeAssociation, data: any) =>
    findFromModel(association.target, isListAssociation(association), applyParser(association.target, data))

  const reduceUpdateData = async (data: any = {}) => {
    const results = {}
    await Promise.all(Object.keys(data).map(async name => {
      const association = model.associations[name]
      if(!model.associations.hasOwnProperty(name)) {
        results[name] = data[name]
        return
      }
      const modelData = await getModelData(association, data[name])
      if(!modelData)
        // tslint:disable-next-line:max-line-length
        throw new Error(`Could not find instances for ${name}, are you sure about request data: ${JSON.stringify(data[name])}`)
      const modelKey = association.otherKey || association.targetKey || association.foreignKey
      results[association.through ? name : association.identifierField] =
        association.through ? modelData.map(({ dataValues }) => dataValues[modelKey]) : modelData[modelKey]
    }))
    return results
  }

  const updateAssociations = (data: any, instances: any[]) =>
    Promise.all(Object.keys(model.associations).map(async name => {
      console.log('updating:', name, data)
      const association = model.associations[name]
      if(!data.hasOwnProperty(name))
        return
      const modelData = await getModelData(association, data[name])
      console.log('update assoc:', modelData)
      return Promise.all(instances.map(result => result[association.accessors.set](modelData)))
    }))

  const resolvers = {
    create: async (_, { data }) => {
      console.log(data)
      const updateData = await reduceUpdateData(data)
      const instance = await model.create(updateData)
      await updateAssociations(data, [instance])
      return instance
    },
    delete: async (_, { where: { where } }) => {
      const deletedItems = await model.findAll({ where })
      await model.destroy({ where })
      return deletedItems
    },
    findMany: async (_, { order, where: { where = {}, include = [] } = {} }) => findAll(include, where, order),
    findOne: async (_, { order, where: { where = {}, include = [] } = {} }) => model.findOne({ include, where, order }),
    update: async (_, { data, where: { where, include } }) => {
      await model.update(data, { where })
      const results = await model.findAll({ where })

      await Promise.all(Object.keys(model.associations).map(async name => {
        const association = model.associations[name]
        if(!data.hasOwnProperty(name))
          return
        const modelData = await getModelData(association, data[name])
        return Promise.all(results.map(result => result[association.accessors.set](modelData)))
      }))
      return results
    },
  }
  return resolvers
})
