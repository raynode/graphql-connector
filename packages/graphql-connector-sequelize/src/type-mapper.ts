import { TypeMapper } from '@raynode/graphql-connector'

import {
  GraphQLBoolean,
  GraphQLEnumType,
  GraphQLFloat,
  GraphQLID,
  GraphQLInt,
  GraphQLList,
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLString,
  GraphQLType,
} from 'graphql'

import { capitalize } from 'inflection'
import { mapValues, upperFirst } from 'lodash'

import { DateType, JSONType, UploadType } from './types'

import { DataTypeAbstract, DataTypes, guards, RangeSubTypes } from './type-guards'

// Map of special characters
const specialCharsMap = new Map([['¼', 'frac14'], ['½', 'frac12'], ['¾', 'frac34']])

const sanitizeEnumValue = (value: string) => {
  return value
    .trim()
    .replace(/([^_a-zA-Z0-9])/g, (_, p) => specialCharsMap.get(p) || ' ')
    .split(' ')
    .map((v, i) => (i ? upperFirst(v) : v))
    .join('')
    .replace(/(^\d)/, '_$1')
}

export const toGraphQLScalar = (sequelizeType: DataTypeAbstract): GraphQLScalarType => {
  if (guards.isBoolean(sequelizeType)) return GraphQLBoolean

  if (guards.isFloatType(sequelizeType)) return GraphQLFloat

  if (guards.isDateType(sequelizeType)) return DateType

  if (guards.isStringType(sequelizeType)) return GraphQLString

  if (guards.isUUID(sequelizeType)) return GraphQLID

  if (guards.isIntegerType(sequelizeType)) return GraphQLInt

  console.log(sequelizeType)
  throw new Error('Unkown Scalar Type found')
}

export const rangeSubTypeToGraphQL = (subtype: RangeSubTypes) => {
  switch (subtype.toLowerCase()) {
    case 'integer':
      return GraphQLInt
    case 'bigint':
      return GraphQLString
    case 'decimal':
      return GraphQLString
    case 'dateonly':
      return DateType
    case 'date':
      return DateType
    case 'datenotz':
      return DateType
    case 'blob':
      return UploadType
    default:
      throw new Error(`Unkown subtype: ${subtype}`)
  }
}

export const typeMapper: TypeMapper<DataTypes, any> = (attribute, model) => {
  const sequelizeType = attribute.type

  if (!guards.isAbstract(sequelizeType)) {
    console.error(sequelizeType)
    throw new Error('Input is not of type abstract-sequelize-type')
  }

  if (guards.isScalarType(sequelizeType)) return toGraphQLScalar(sequelizeType)

  if (guards.isArray(sequelizeType)) return new GraphQLList(toGraphQLScalar(sequelizeType.type))

  if (guards.isEnum(sequelizeType)) {
    const values = mapValues(sequelizeType.values.map(sanitizeEnumValue), value => ({ value }))
    return new GraphQLEnumType({ name: `${model.name.toString()}${capitalize(attribute.name)}EnumType`, values })
  }

  if (guards.isVirtual(sequelizeType))
    return sequelizeType.returnType ? toGraphQLScalar(sequelizeType.returnType) : GraphQLString

  if (guards.isJson(sequelizeType)) return JSONType

  if (guards.isBlob(sequelizeType)) return UploadType

  if (guards.isRange(sequelizeType)) {
    const type = rangeSubTypeToGraphQL(sequelizeType._subtype)
    const name = `${capitalize(sequelizeType._subtype)}RangeType`
    return new GraphQLObjectType({
      name,
      fields: {
        includeLowerBound: { type: GraphQLBoolean },
        includeUpperBound: { type: GraphQLBoolean },
        lowerBound: { type },
        upperBound: { type },
      },
    })
  }

  console.error(sequelizeType)
  throw new Error(`Unable to convert ${sequelizeType.key}::${sequelizeType.toSql()} to a GraphQL type`)
}
