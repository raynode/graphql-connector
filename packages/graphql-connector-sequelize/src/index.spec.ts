import { createBaseSchemaGenerator, createSchema } from '@raynode/graphql-connector'
import { graphql, GraphQLSchema, printSchema } from 'graphql'
import * as Sequelize from 'sequelize'
import { configuration } from './index'
import { modelMapper } from './model-mapper'
import { typeMapper } from './type-mapper'

import { initialize, models, uuidv4 } from './tests/sample-models'

describe('model-mapper', () => {
  it('should find all attributes from a model', () => {
    const result = modelMapper('User', models.User)
    expect(result.attributes).toHaveProperty('id')
    expect(result.attributes).toHaveProperty('state')
    expect(result.attributes).toHaveProperty('nickname')
    expect(result.attributes).toHaveProperty('name')
    expect(result.attributes).toHaveProperty('email')
    expect(result.attributes).toHaveProperty('createdAt')
    expect(result.attributes).toHaveProperty('updatedAt')
    expect(Object.keys(result.attributes)).toHaveLength(7)
  })
})

describe('schema', () => {
  let schema: GraphQLSchema
  const runQuery = async (query: string) => graphql(schema, query, null, null)

  beforeAll(() => initialize())

  beforeEach(async () => {
    const baseSchemaGenerator = createBaseSchemaGenerator(configuration)
    const baseSchema = baseSchemaGenerator(models)
    schema = createSchema(baseSchema)
  })

  it('should generate a full schema', () => {
    expect(printSchema(schema)).toMatchSnapshot()
  })

  it('should correctly load the data', async () => {
    const { data } = await runQuery(`{
      Users { nodes {
        id
        state
        nickname
        name
        email
        createdAt
        updatedAt
      } }
    }`)
    expect(data.Users.nodes).toMatchSnapshot()
  })

  it('should create a new user', async () => {
    const { data } = await runQuery(`
      mutation {
        newUser: createUser(data: {
          name: "Jack"
          nickname: "Jacky"
          email: "jack@example.com"
        }) {
          id
          name
          nickname
          email
        }
      }
    `)
    expect(data).toMatchSnapshot()
  })

  it('should now find Georg as well', async () => {
    const { data } = await runQuery(`{
      Users { nodes {
        id
        state
        nickname
        name
        email
      } }
    }`)
    expect(data.Users.nodes).toMatchSnapshot()
  })

  it('should find only Georg', async () => {
    const { data } = await runQuery(`{
      georg: User(where: {
        name: "Jack"
      }) {
        id
        state
        nickname
        name
        email
      }
    }`)
    expect(data.georg).toMatchSnapshot()
  })

  it('should find the users in ascending order of id', async () => {
    const { data } = await runQuery(`{
      Users(order: id_ASC) { nodes { id name } }
    }`)
    expect(data.Users.nodes[0].id).toEqual(uuidv4(1))
    expect(data).toMatchSnapshot()
  })

  it('should find the users in descending order of id', async () => {
    const { data } = await runQuery(`{
      Users(order: id_DESC) { nodes { id name } }
    }`)
    expect(data.Users.nodes[0].id).toEqual(uuidv4(101))
    expect(data).toMatchSnapshot()
  })

  it('should find all posts and their authors', async () => {
    const { data } = await runQuery(`{
      Posts { nodes {
        title
        User {
          name
        }
      }}
    }`)
    expect(data).toMatchSnapshot()
  })

  it('should find everybody except the admin', async () => {
    const { data } = await runQuery(`{
      Users(where: { NOT: {
        state: admin
      }}) { nodes {
        name
        state
      }}
    }`)
    expect(data).toMatchSnapshot()
  })

  it('should find all users that have written a post', async () => {
    const { data } = await runQuery(`{
      Users(where: {
        Posts_some: { NOT: { id: "" }}
      }) { nodes {
        name
        Posts { nodes {
          title
        }}
      }}
    }`)
    expect(data).toMatchSnapshot()
  })

  it('should create a post', async () => {
    const { data, errors } = await runQuery(`mutation {
      createPost(data: {
        title: "new post"
        User: {
          id: "${uuidv4(4)}",
        }
      }) {
        id
        User { name }
      }
    }`)
    expect(data.createPost.id).toEqual(uuidv4(102))
    expect(data).toMatchSnapshot()
  })

  it('should update a post', async () => {
    const { data } = await runQuery(`mutation {
      updatePost(data: {
        title: "This is my new Post"
      }, where: {
        id: "${uuidv4(102)}"
      }) {
        id
        title
      }
    }`)
    expect(data).toMatchSnapshot()
  })

  it('should find an updated post', async () => {
    const { data } = await runQuery(`{
      Post(where: {
        id: "${uuidv4(102)}"
      }) {
        title
      }
    }`)
    expect(data.Post.title).toEqual('This is my new Post')
  })

  it('should change the user of the post', async () => {
    const { data } = await runQuery(`mutation {
      updatePost(where: {
        id: "${uuidv4(102)}"
      }, data: {
        User: {
          id: "${uuidv4(1)}"
        }
      }) {
        title
        User { name }
      }
    }`)
    expect(data.updatePost[0].User.name).toEqual('Admin')
    expect(data).toMatchSnapshot()
  })

  it('should delete the post', async () => {
    const { data } = await runQuery(`mutation {
      deletePosts(where: {
        id: "${uuidv4(102)}"
      }) {
        title
      }
    }`)
    expect(data.deletePosts[0].title).toEqual('This is my new Post')
  })

  it('should not find the post', async () => {
    const { data } = await runQuery(`{
      Post(where: {
        id: "${uuidv4(102)}"
      }) {
        title
      }
    }`)
    expect(data.Post).toBeNull()
  })

  it('should create a bookmark for Paulchen', async () => {
    const { data } = await runQuery(`mutation {
      createLink(data: {
        url: "https://example.com"
        title: "Examples"
        user: {
          nickname: "Paulchen"
        }
      }) {
        id
        title
        url
        user { name nickname }
      }
    }`)
    expect(data).toMatchSnapshot()
  })

  it('should create a link with a tag', async () => {
    const { data } = await runQuery(`mutation {
      createLink(data: {
        url: "https://example.com"
        title: "Mehr Examples"
        tags: {
          tag: "Nice Pages"
        }
      }) {
        id
        title
        url
        tags { nodes { tag } }
      }
    }`)
    expect(data.createLink.tags.nodes).toHaveLength(1)
  })

  it('should list all links', async () => {
    const { data, errors } = await runQuery(`{
      Links { nodes {
        id
        title
        url
        user { name nickname }
        tags { nodes { tag } }
      } }
    }`)
    if(errors) return console.log(errors)
    console.log(data.Links.nodes)
  })

  it('should list all tags', async () => {
    const { data, errors } = await runQuery(`{
      Tags(where: { tag: "Nice Pages" }) { nodes {
        tag
      } }
    }`)
    if(errors) return console.log(errors)
    console.log(data.Tags.nodes)
  })
})
