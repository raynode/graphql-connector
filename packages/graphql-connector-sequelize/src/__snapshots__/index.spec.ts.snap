// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema should change the user of the post 1`] = `
Object {
  "updatePost": Array [
    Object {
      "User": Object {
        "name": "Admin",
      },
      "title": "This is my new Post",
    },
  ],
}
`;

exports[`schema should correctly load the data 1`] = `
Array [
  Object {
    "createdAt": "2018-11-09T16:00:00.000Z",
    "email": "admin@example.com",
    "id": "1",
    "name": "Admin",
    "nickname": "Admin",
    "state": "admin",
    "updatedAt": "2018-11-09T16:00:00.000Z",
  },
  Object {
    "createdAt": "2018-11-09T16:30:00.000Z",
    "email": "georg@example.com",
    "id": "2",
    "name": "Georg",
    "nickname": "Mr.G!",
    "state": "member",
    "updatedAt": "2018-11-09T16:30:00.000Z",
  },
  Object {
    "createdAt": "2018-11-09T16:40:00.000Z",
    "email": "paul@example.com",
    "id": "3",
    "name": "Paul",
    "nickname": "Paulchen",
    "state": "guest",
    "updatedAt": "2018-11-09T16:40:00.000Z",
  },
  Object {
    "createdAt": "2018-11-09T16:50:00.000Z",
    "email": "frank@example.com",
    "id": "4",
    "name": "Frank",
    "nickname": "Effy",
    "state": "member",
    "updatedAt": "2018-11-09T16:50:00.000Z",
  },
]
`;

exports[`schema should create a bookmark for Paulchen 1`] = `
Object {
  "createLink": Object {
    "id": "105",
    "title": "Examples",
    "url": "https://example.com",
    "user": Object {
      "name": "Paul",
      "nickname": "Paulchen",
    },
  },
}
`;

exports[`schema should create a new user 1`] = `
Object {
  "newUser": Object {
    "email": "jack@example.com",
    "id": "101",
    "name": "Jack",
    "nickname": "Jacky",
  },
}
`;

exports[`schema should create a post 1`] = `
Object {
  "createPost": Object {
    "User": Object {
      "name": "Frank",
    },
    "id": "102",
  },
}
`;

exports[`schema should find all posts and their authors 1`] = `
Object {
  "Posts": Object {
    "nodes": Array [
      Object {
        "User": Object {
          "name": "Admin",
        },
        "title": "This is a Post by the Admin",
      },
      Object {
        "User": Object {
          "name": "Paul",
        },
        "title": "This is by Paulchen",
      },
    ],
  },
}
`;

exports[`schema should find all users that have written a post 1`] = `
Object {
  "Users": Object {
    "nodes": Array [
      Object {
        "Posts": Object {
          "nodes": Array [
            Object {
              "title": "This is a Post by the Admin",
            },
          ],
        },
        "name": "Admin",
      },
      Object {
        "Posts": Object {
          "nodes": Array [
            Object {
              "title": "This is by Paulchen",
            },
          ],
        },
        "name": "Paul",
      },
    ],
  },
}
`;

exports[`schema should find everybody except the admin 1`] = `
Object {
  "Users": Object {
    "nodes": Array [
      Object {
        "name": "Georg",
        "state": "member",
      },
      Object {
        "name": "Paul",
        "state": "guest",
      },
      Object {
        "name": "Frank",
        "state": "member",
      },
      Object {
        "name": "Jack",
        "state": "guest",
      },
    ],
  },
}
`;

exports[`schema should find only Georg 1`] = `
Object {
  "email": "jack@example.com",
  "id": "101",
  "name": "Jack",
  "nickname": "Jacky",
  "state": "guest",
}
`;

exports[`schema should find the users in ascending order of id 1`] = `
Object {
  "Users": Object {
    "nodes": Array [
      Object {
        "id": "1",
        "name": "Admin",
      },
      Object {
        "id": "2",
        "name": "Georg",
      },
      Object {
        "id": "3",
        "name": "Paul",
      },
      Object {
        "id": "4",
        "name": "Frank",
      },
      Object {
        "id": "101",
        "name": "Jack",
      },
    ],
  },
}
`;

exports[`schema should find the users in descending order of id 1`] = `
Object {
  "Users": Object {
    "nodes": Array [
      Object {
        "id": "101",
        "name": "Jack",
      },
      Object {
        "id": "4",
        "name": "Frank",
      },
      Object {
        "id": "3",
        "name": "Paul",
      },
      Object {
        "id": "2",
        "name": "Georg",
      },
      Object {
        "id": "1",
        "name": "Admin",
      },
    ],
  },
}
`;

exports[`schema should generate a full schema 1`] = `
"input CreateLinkData {
  id: ID
  title: String
  url: String
  createdAt: Date
  updatedAt: Date
  user: UserWhere
  tags: TagWhere
}

input CreatePostData {
  id: ID
  title: String
  createdAt: Date
  updatedAt: Date
  User: UserWhere
  tags: TagWhere
}

input CreateTagData {
  id: ID
  tag: String
  createdAt: Date
  updatedAt: Date
  Users: UserWhere
  Links: LinkWhere
  Posts: PostWhere
}

input CreateUserData {
  id: ID
  state: UserStateEnumType
  nickname: String
  name: String
  email: String
  createdAt: Date
  updatedAt: Date
  Posts: PostWhere
  bookmarks: LinkWhere
  tags: TagWhere
}

\\"\\"\\"
A special custom Scalar type for Dates that converts to a ISO formatted string 
\\"\\"\\"
scalar Date

type Link implements Node {
  id: ID!
  title: String!
  url: String!
  createdAt: Date!
  updatedAt: Date!
  user: User
  tags: Tags!
}

input LinkFilter {
  AND: [LinkFilter!]
  OR: [LinkFilter!]
  NOT: LinkFilter
  id: ID
  id_in: [ID!]
  title: String
  title_in: [String!]
  title_like: [String!]
  title_startsWith: [String!]
  title_endsWith: [String!]
  url: String
  url_in: [String!]
  url_like: [String!]
  url_startsWith: [String!]
  url_endsWith: [String!]
}

type Links implements List {
  nodes: [Link!]!
  page: Page!
}

enum LinkSortOrder {
  id_ASC
  id_DESC
  title_ASC
  title_DESC
  url_ASC
  url_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  user_ASC
  user_DESC
  tags_ASC
  tags_DESC
}

input LinkWhere {
  AND: [LinkFilter!]
  OR: [LinkFilter!]
  NOT: LinkFilter
  id: ID
  id_in: [ID!]
  title: String
  title_in: [String!]
  title_like: [String!]
  title_startsWith: [String!]
  title_endsWith: [String!]
  url: String
  url_in: [String!]
  url_like: [String!]
  url_startsWith: [String!]
  url_endsWith: [String!]
  user_empty: UserFilter
  user_some: UserFilter
  tags_empty: TagFilter
  tags_some: TagFilter
}

interface List {
  nodes: [Node!]!
  page: Page!
}

type Mutation {
  createUser(data: CreateUserData): User
  updateUser(data: UpdateUserData, where: UserWhere): [User!]!
  deleteUsers(where: UserWhere, order: UserSortOrder): [User!]!
  createLink(data: CreateLinkData): Link
  updateLink(data: UpdateLinkData, where: LinkWhere): [Link!]!
  deleteLinks(where: LinkWhere, order: LinkSortOrder): [Link!]!
  createPost(data: CreatePostData): Post
  updatePost(data: UpdatePostData, where: PostWhere): [Post!]!
  deletePosts(where: PostWhere, order: PostSortOrder): [Post!]!
  createTag(data: CreateTagData): Tag
  updateTag(data: UpdateTagData, where: TagWhere): [Tag!]!
  deleteTags(where: TagWhere, order: TagSortOrder): [Tag!]!
}

interface Node {
  id: ID
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Post implements Node {
  id: ID!
  title: String!
  createdAt: Date!
  updatedAt: Date!
  User: User
  tags: Tags!
}

input PostFilter {
  AND: [PostFilter!]
  OR: [PostFilter!]
  NOT: PostFilter
  id: ID
  id_in: [ID!]
  title: String
  title_in: [String!]
  title_like: [String!]
  title_startsWith: [String!]
  title_endsWith: [String!]
}

type Posts implements List {
  nodes: [Post!]!
  page: Page!
}

enum PostSortOrder {
  id_ASC
  id_DESC
  title_ASC
  title_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  User_ASC
  User_DESC
  tags_ASC
  tags_DESC
}

input PostWhere {
  AND: [PostFilter!]
  OR: [PostFilter!]
  NOT: PostFilter
  id: ID
  id_in: [ID!]
  title: String
  title_in: [String!]
  title_like: [String!]
  title_startsWith: [String!]
  title_endsWith: [String!]
  User_empty: UserFilter
  User_some: UserFilter
  tags_empty: TagFilter
  tags_some: TagFilter
}

type Query {
  User(where: UserWhere, order: UserSortOrder): User
  Users(where: UserWhere, order: UserSortOrder): Users!
  Link(where: LinkWhere, order: LinkSortOrder): Link
  Links(where: LinkWhere, order: LinkSortOrder): Links!
  Post(where: PostWhere, order: PostSortOrder): Post
  Posts(where: PostWhere, order: PostSortOrder): Posts!
  Tag(where: TagWhere, order: TagSortOrder): Tag
  Tags(where: TagWhere, order: TagSortOrder): Tags!
}

type Subscription {
  onCreateUser: User!
  onDeleteUsers: [User!]!
  onUpdateUser: User
  onCreateLink: Link!
  onDeleteLinks: [Link!]!
  onUpdateLink: Link
  onCreatePost: Post!
  onDeletePosts: [Post!]!
  onUpdatePost: Post
  onCreateTag: Tag!
  onDeleteTags: [Tag!]!
  onUpdateTag: Tag
}

type Tag implements Node {
  id: ID!
  tag: String!
  createdAt: Date!
  updatedAt: Date!
  Users: Users!
  Links: Links!
  Posts: Posts!
}

input TagFilter {
  AND: [TagFilter!]
  OR: [TagFilter!]
  NOT: TagFilter
  id: ID
  id_in: [ID!]
  tag: String
  tag_in: [String!]
  tag_like: [String!]
  tag_startsWith: [String!]
  tag_endsWith: [String!]
}

type Tags implements List {
  nodes: [Tag!]!
  page: Page!
}

enum TagSortOrder {
  id_ASC
  id_DESC
  tag_ASC
  tag_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  TagLinks_ASC
  TagLinks_DESC
  Users_ASC
  Users_DESC
  Links_ASC
  Links_DESC
  Posts_ASC
  Posts_DESC
}

input TagWhere {
  AND: [TagFilter!]
  OR: [TagFilter!]
  NOT: TagFilter
  id: ID
  id_in: [ID!]
  tag: String
  tag_in: [String!]
  tag_like: [String!]
  tag_startsWith: [String!]
  tag_endsWith: [String!]
  Users_empty: UserFilter
  Users_some: UserFilter
  Links_empty: LinkFilter
  Links_some: LinkFilter
  Posts_empty: PostFilter
  Posts_some: PostFilter
}

input UpdateLinkData {
  id: ID
  title: String
  url: String
  createdAt: Date
  updatedAt: Date
  user: LinkWhere
  tags: LinkWhere
}

input UpdatePostData {
  id: ID
  title: String
  createdAt: Date
  updatedAt: Date
  User: PostWhere
  tags: PostWhere
}

input UpdateTagData {
  id: ID
  tag: String
  createdAt: Date
  updatedAt: Date
  Users: TagWhere
  Links: TagWhere
  Posts: TagWhere
}

input UpdateUserData {
  id: ID
  state: UserStateEnumType
  nickname: String
  name: String
  email: String
  createdAt: Date
  updatedAt: Date
  Posts: UserWhere
  bookmarks: UserWhere
  tags: UserWhere
}

type User implements Node {
  id: ID!
  state: UserStateEnumType
  nickname: String
  name: String!
  email: String!
  createdAt: Date!
  updatedAt: Date!
  Posts: Posts!
  bookmarks: Links!
  tags: Tags!
}

input UserFilter {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  id: ID
  id_in: [ID!]
  state: UserStateEnumType
  state_in: [UserStateEnumType!]
  nickname: String
  nickname_in: [String!]
  nickname_like: [String!]
  nickname_startsWith: [String!]
  nickname_endsWith: [String!]
  name: String
  name_in: [String!]
  name_like: [String!]
  name_startsWith: [String!]
  name_endsWith: [String!]
  email: String
  email_in: [String!]
  email_like: [String!]
  email_startsWith: [String!]
  email_endsWith: [String!]
}

type Users implements List {
  nodes: [User!]!
  page: Page!
}

enum UserSortOrder {
  id_ASC
  id_DESC
  state_ASC
  state_DESC
  nickname_ASC
  nickname_DESC
  name_ASC
  name_DESC
  email_ASC
  email_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  Posts_ASC
  Posts_DESC
  bookmarks_ASC
  bookmarks_DESC
  tags_ASC
  tags_DESC
}

enum UserStateEnumType {
  admin
  member
  guest
}

input UserWhere {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  id: ID
  id_in: [ID!]
  state: UserStateEnumType
  state_in: [UserStateEnumType!]
  nickname: String
  nickname_in: [String!]
  nickname_like: [String!]
  nickname_startsWith: [String!]
  nickname_endsWith: [String!]
  name: String
  name_in: [String!]
  name_like: [String!]
  name_startsWith: [String!]
  name_endsWith: [String!]
  email: String
  email_in: [String!]
  email_like: [String!]
  email_startsWith: [String!]
  email_endsWith: [String!]
  Posts_empty: PostFilter
  Posts_some: PostFilter
  bookmarks_empty: LinkFilter
  bookmarks_some: LinkFilter
  tags_empty: TagFilter
  tags_some: TagFilter
}
"
`;

exports[`schema should now find Georg as well 1`] = `
Array [
  Object {
    "email": "admin@example.com",
    "id": "1",
    "name": "Admin",
    "nickname": "Admin",
    "state": "admin",
  },
  Object {
    "email": "georg@example.com",
    "id": "2",
    "name": "Georg",
    "nickname": "Mr.G!",
    "state": "member",
  },
  Object {
    "email": "paul@example.com",
    "id": "3",
    "name": "Paul",
    "nickname": "Paulchen",
    "state": "guest",
  },
  Object {
    "email": "frank@example.com",
    "id": "4",
    "name": "Frank",
    "nickname": "Effy",
    "state": "member",
  },
  Object {
    "email": "jack@example.com",
    "id": "101",
    "name": "Jack",
    "nickname": "Jacky",
    "state": "guest",
  },
]
`;

exports[`schema should update a post 1`] = `
Object {
  "updatePost": Array [
    Object {
      "id": "102",
      "title": "This is my new Post",
    },
  ],
}
`;
